src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,128 @@
+  killState: false,
+  setKillState: () => {},
+  setPasscode: () => {},
+  passcode: null,
+  checkFlag: false,
+  setCheckFlag: () => {},
+  killStateActive: false,
+  setKillStateActive: () => {},
+};
+
+export const SecurityContext = React.createContext<SecuritContextType>(
+  DEFAULT_CONTEXT
+);
+
+export const SecurityProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [killState, setKillState] = React.useState(false);
+  const [killStateActive, setKillStateActive] = React.useState(false);
+  const [passcode, setPasscode] = React.useState<string | null>(null);
+  const [checkFlag, setCheckFlag] = React.useState(false);
+  const firstRender = useRef(true);
+
+  useEffect(() => {
+    if (firstRender.current) {
+      AsyncStorage.getItem(PASSWORD_KEY, (err, val) => {
+        if (!!val) {
+          setPasscode(val);
+          setCheckFlag(true);
+          AsyncStorage.getItem(ASYNC_KILL_KEY, (err, val) => {
+            if (!!val) setKillStateActive(val === "true");
+          });
+        } else {
+          firstRender.current = false;
+        }
+      });
+    }
+  }, []);
+
+  useEffect(() => {
+    //Checks if the pass is just being set from initial render
+    if (firstRender.current) {
+      firstRender.current = false;
+      return;
+    }
+
+    if (!passcode) {
+      setKillState(false);
+      setKillStateActive(false);
+      setCheckFlag(false);
+      AsyncStorage.setItem(PASSWORD_KEY, "");
+    } else {
+      AsyncStorage.setItem(PASSWORD_KEY, passcode);
+    }
+  }, [passcode]);
+
+  useEffect(() => {
+    AsyncStorage.setItem(ASYNC_KILL_KEY, JSON.stringify(killStateActive));
+  }, [killStateActive]);
+
+  //checks that the user has a passcode set before setting kill state
+  const setKillStateActiveWithCheck = React.useCallback(
+    (newState: boolean) => {
+      if (!passcode && newState) {
+        throw new Error("Cannot set kill state without a passcode");
+      }
+      setKillStateActive(newState);
+    },
+    [setKillStateActive, passcode]
+  );
+
+  //validates passcode
+  const setPasscodeWithCheck = (passcode: string | null) => {
+    if (!passcode) {
+      setPasscode(null);
+    } else {
+      if (!validPasscode(passcode)) {
+        throw new Error("passcode not valid");
+      } else {
+        setPasscode(passcode);
+      }
+    }
+  };
+
+  const contextValues: SecuritContextType = React.useMemo(() => {
+    return {
+      killState,
+      setKillState,
+      setPasscode: setPasscodeWithCheck,
+      checkFlag,
+      setCheckFlag,
+      passcode,
+      killStateActive,
+      setKillStateActive: setKillStateActiveWithCheck,
+    };
+  }, [killState, passcode, checkFlag, killStateActive]);
+
+  return (
+    <SecurityContext.Provider value={contextValues}>
+      {children}
+    </SecurityContext.Provider>
+  );
+};
+
+function validPasscode(passcode: string): boolean {
+  return passcode.length === 5 && !isNaN(parseInt(passcode));
```

Comment:
as a precaution, pass `10` as the radix in `parseInt` i.e. `parseInt(passcode, 10))` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#description